### PR TITLE
build: add support for go 1.22 (backport #13738)

### PR DESCRIPTION
**Pull request by smoshiur1237**
Backport : #13738 
It needs to support go 1.22 to resolve vulnerabilities in go 1.20

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by smoshiur1237** [comment]
@travisn @parth-gr  please check, if we can take it in.

**Comment by subhamkrai** [comment]
@smoshiur1237 Please check and fix the CI error

**Comment by smoshiur1237** [comment]
> @smoshiur1237 Please check and fix the CI error

@subhamkrai would you please give direction where should I need to do changes. I am not familiar with the CI.

**Comment by subhamkrai** [comment]
@smoshiur1237 can you rebase your PR with latest master and push again?

File: Makefile [hunk 1]
```diff
@@ -32,7 +32,7 @@ all: build
 
 # Controller-gen version
 # f284e2e8... is master ahead of v0.5.0 which has ability to generate embedded objectmeta in CRDs
-CONTROLLER_GEN_VERSION=v0.11.3
+CONTROLLER_GEN_VERSION=v0.14.0
```
Comment (travisn) [inline Makefile:35]: Is this change actually needed? Or can we revert this? If we can avoid the regen of the crds, it will simplify this PR a lot.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -77,7 +77,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        go-version: ["1.21","1.20"]
+        go-version: ["1.21", "1.22"]
```
Comment (travisn) [inline .github/workflows/build.yml:80]: We don't want to remove any support for 1.20.
```suggestion
        go-version: ["1.20", "1.21", "1.22"]
```

**Comment by travisn** [comment]
Looks like the CI is failing with old unsupported actions:
```
Error: This request has been automatically failed because it uses a deprecated version of `actions/upload-artifact: v2`. Learn more: https://github.blog/changelog/2024-02-13-deprecation-notice-v1-and-v2-of-the-artifact-actions/
```

This is why it is difficult to support the older branches, because we don't maintain the CI actions. So it will be difficult to get a green CI to merge anything to the 1.13 branch. See also the [support policy](https://rook.io/docs/rook/latest/Getting-Started/release-cycle/).

If we can restrict the change to only adding go 1.22 to the allowed versions (only your change in [.github/workflows/build.yml](https://github.com/rook/rook/pull/14809/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721)), and no other code changes, I'm ok to merge it even with a broken CI. But we cannot merge any code changes (including crds.yaml) if the CI is not green.

I would highly recommend you build from a fork so you can fully control build issues like this.

> @smoshiur1237 can you rebase your PR with latest master and push again?

The CI fixes are in master, but we can't rebase against 1.13 branch, it would require specific backports to avoid pulling in the entire set of changes in master.

**Review by travisn — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 2 above)
Comment (travisn) [reply .github/workflows/build.yml:80]: Actually, instead of modifying this, you'll need to modify the GO_SUPPORTED_VERSIONS [here](https://github.com/rook/rook/blob/release-1.13/build/makelib/golang.mk#L51).

**Review by travisn — CHANGES_REQUESTED** [review]

File: build/makelib/golang.mk [hunk 3]
```diff
@@ -48,7 +48,7 @@ GO_TEST_FLAGS ?=
 # ====================================================================================
 # Setup go environment
 
-GO_SUPPORTED_VERSIONS ?= 1.21|1.20
+GO_SUPPORTED_VERSIONS ?= 1.21|1.22
```
Comment (travisn) [inline build/makelib/golang.mk:51]: need to keep 1.20
```suggestion
GO_SUPPORTED_VERSIONS ?= 1.20|1.21|1.22
```

**Review by travisn — APPROVED** [review]
I confirmed locally that this will allow building with go 1.22, so we can go ahead and merge. 

However, Rook does not have plans to release based on this new version of golang, due to the [support policy](https://rook.io/docs/rook/latest/Getting-Started/release-cycle/). But the two changes that would be required for a release would at least include:
1. Update the CI actions to confirm a green build (these changes are already made in 1.14 and newer branches, but not maintained in 1.13)
2. Set the Go version in the release github action to use go 1.22, see [here](https://github.com/rook/rook/blob/release-1.13/.github/workflows/push-build.yaml#L28-L30).
